jaegertracing/jaegertracing.go

```diff
@@ -128,82 +129,101 @@ func TraceWithConfig(config TraceConfig) echo.MiddlewareFunc {
 			}
+			defer sp.Finish()
 
 			ext.HTTPMethod.Set(sp, req.Method)
 			ext.HTTPUrl.Set(sp, req.URL.String())
 			ext.Component.Set(sp, config.ComponentName)
+			sp.SetTag("client_ip", realIP)
+			sp.SetTag("request_id", requestID)
 
 			// Dump request & response body
-			resBody := new(bytes.Buffer)
+			var respDumper *ResponseDumper
 			if config.IsBodyDump {
 				// request
 				reqBody := []byte{}
-				if c.Request().Body != nil { // Read
+				if c.Request().Body != nil {
 					reqBody, _ = ioutil.ReadAll(c.Request().Body)
-					sp.SetTag("http.req.body", string(reqBody))
+					sp.LogKV("http.req.body", string(reqBody))
 				}
 
-				req.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody)) // Reset
+				req.Body = ioutil.NopCloser(bytes.NewBuffer(reqBody)) // reset original request body
 
 				// response
-				mw := io.MultiWriter(c.Response().Writer, resBody)
-				writer := &bodyDumpResponseWriter{Writer: mw, ResponseWriter: c.Response().Writer}
-				c.Response().Writer = writer
+				respDumper = NewResponseDumper(c.Response())
+				c.Response().Writer = respDumper
 			}
 
+			// setup request context - add opentracing span
 			req = req.WithContext(opentracing.ContextWithSpan(req.Context(), sp))
 			c.SetRequest(req)
 
-			var err error
-			defer func() {
-				committed := c.Response().Committed
-				status := c.Response().Status
-
-				if err != nil {
-					var httpError *echo.HTTPError
-					if errors.As(err, &httpError) {
-						if httpError.Code != 0 {
-							status = httpError.Code
-						}
-						sp.SetTag("error.message", httpError.Message)
-					} else {
-						sp.SetTag("error.message", err.Error())
-					}
-					if status == http.StatusOK {
-						// this is ugly workaround for cases when httpError.code == 0 or error was not httpError and status
-						// in request was 200 (OK). In these cases replace status with something that represents an error
-						// it could be that error handlers or middlewares up in chain will output different status code to
-						// client. but at least we send something better than 200 to jaeger
-						status = http.StatusInternalServerError
-					}
-				}
+			// call next middleware / controller
+			err = next(c)
+			if err != nil {
+				c.Error(err) // call custom registered error handler
+			}
 
-				ext.HTTPStatusCode.Set(sp, uint16(status))
-				if status >= http.StatusInternalServerError || !committed {
-					ext.Error.Set(sp, true)
-				}
+			status := c.Response().Status
+			ext.HTTPStatusCode.Set(sp, uint16(status))
 
-				// Dump response body
-				if config.IsBodyDump {
-					sp.SetTag("http.resp.body", resBody.String())
-				}
+			if err != nil {
+				logError(sp, err)
+			}
 
-				sp.Finish()
-			}()
-			err = next(c)
-			return err
+			// Dump response body
+			if config.IsBodyDump {
+				sp.LogKV("http.resp.body", respDumper.GetResponse())
+			}
+
+			// echo request_id back in response
+			c.Response().Header().Set(echo.HeaderXRequestID, requestID)
```

Comment:
This is actually something that the RequestID middleware handles.
Is there es specific reason it is required for jaegertracing?

Reply:
No there is no specific reason for doing it in jaegertracing.